### ISSUE TITLE
Arduino ESP32 Nano: Uploading multiple files is not reliable 

### ISSUE THREAD

**Issue by JohannesMaierhofer** [micropython/micropython#12474]
Uploading multiple .py files using Thonny and the newest Micropython firmware v1.20.0-475-g174bb28d8 on 2023-09-14; Arduino Nano ESP32 with ESP32S3 is not stable. Often the files end up being corrupt on the flash of the esp32.
Also uploading only single files that changed is not possible. All other files become corrupt. 

It also happens when uploading the files using the Arduino uPython Lab software.

**Comment by jimmo** [comment]
@JohannesMaierhofer Neither Thonny or Arduino uPython Lab are maintained by the micropython team.

Do you see the same issues with `mpremote`? https://docs.micropython.org/en/latest/reference/mpremote.html

You mentioned the issue is with "the newest MicroPython firmware". Do you mean it was working before, or is this the first version you've used.

I notice the Nano ESP32 uses FAT not LittleFS. I will try setting up the Nano filesystem layout on a regular ESP32S3 and see if it works for me. Maybe cc @pillo79 or @iabdalkader have thoughts?

**Comment by jimmo** [comment]
I don't have an Arduino ESP32 Nano, but I flashed the Nano firmware onto a similar esp32s3 board, and can't see any issues with the filesystem with mpremote.

FYI @pillo79 `make BOARD=ARDUINO_NANO_ESP32 deploy` doesn't work because it's trying to flash the `factory` partition, where it should be flashing `app0`. I _think_ there just shouldn't be the `factory` line in the board's partitions.csv file?

**Comment by pillo79** [comment]
Thanks for the heads-up @jimmo.

> FYI @pillo79 `make BOARD=ARDUINO_NANO_ESP32 deploy` doesn't work because it's trying to flash the `factory` partition, where it should be flashing `app0`. I _think_ there just shouldn't be the `factory` line in the board's partitions.csv file?

That would be an immediate fix, but would also hide the partition from apps (including uPython) looking for it. We need that to implement recovery from broken uploads, so it's not really a viable option. 

From my first glance, it looks like `idf.py` always [uses the "default partition"](https://github.com/espressif/esp-idf/blob/master/components/esptool_py/CMakeLists.txt#L6C5-L12C5) when deciding where to flash, and [`factory` overrides `app`](https://github.com/espressif/esp-idf/blob/master/components/partition_table/parttool.py#L141-L146) in the search (as expected). I don't see a way to change this either... :raised_eyebrow: 

I will ask Espressif for guidance and let you know!

**Comment by JohannesMaierhofer** [comment]
Hi, I used mpremote today and faced the same issue. I copied the files either as batch or one by one. In the end some files are not in their directory, others are missing and again others are corrupt. 
Everything works well, if I have only 1 main.py file. 
I didn't try older firmwares yet. I am on macOS if this has any influence?

For curiosity I tried an Arduino RP2040 board and every method worked perfectly.

**Comment by pillo79** [comment]
@JohannesMaierhofer thanks for the feedback. I am investigating this and can replicate on a Nano ESP32 - it appears to be an issue with the FAT file system handling in Micropython. I will update this thread with my findings.

**Comment by pillo79** [comment]
Found out the root cause - the FAT was getting corrupted due to an issue in the implementation of partial writes in the `Partition.writeblocks()` function. 

FAT needs to update only areas of a hardware Flash block, but that is not possible at hardware level, so the implementation of `writeblocks` reads, erases and rewrites the block with new contents. There was a mistake in this logic that resulted in data being "offset" when a partial write was performed. I've opened a PR with the one-liner that fixes this.

**Comment by JohannesMaierhofer** [comment]
yes, that's exactly what the issue looked like from the outside. Wow great, thanks for your effort!!
